File: Question/40_Combination_Sum2.cpp
class Solution {
public:
    vector<vector<int>> combinationSum2(vector<int>& candidates, int target) {
        vector<vector<int>> ans;
        vector<int> cur;
        sort(candidates.begin(), candidates.end());
        dfs(candidates, target, 0, 0, cur, ans);
        return ans;
    }
    void dfs(vector<int>& candidates, int target, int loc, int cur_sum, vector<int>& cur, vector<vector<int>>& ans) {
        if(cur_sum > target) return;
        if(cur_sum == target) {
            ans.push_back(cur);
            return;
        }
        for(int i = loc; i < candidates.size(); ++i) {
            if(i > loc && candidates[i] == candidates[i - 1]) continue;
            cur.push_back(candidates[i]);
            dfs(candidates, target, i + 1, cur_sum + candidates[i], cur, ans);
            cur.pop_back();
        }
    }
};

File: Question_Java/40_Combination_Sum2.java
class Solution {
    public List<List<Integer>> combinationSum2(int[] candidates, int target) {
        Arrays.sort(candidates);
        List<List<Integer>> ans = new ArrayList<>();
        dfs(candidates, 0, target, new ArrayList<Integer>(), ans);
        return ans;
    }

    private void dfs(int[] candidates, int loc, int target, List<Integer> combination, List<List<Integer>> ans) {
        if (target < 0)
            return;
        if (target == 0) {
            ans.add(new ArrayList<Integer>(combination));
            return;
        }

        for (int i = loc; i < candidates.length; ++i) {
            combination.add(candidates[i]);
            dfs(candidates, i + 1, target - candidates[i], combination, ans);
            combination.remove(combination.size() - 1);
            while (i + 1 < candidates.length && candidates[i + 1] == candidates[i])
                ++i;
        }
    }
}